src/Bicep.Core/TypeSystem/TupleType.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Bicep.Core.TypeSystem;
+
+/// <summary>
+/// Represents an array with a fixed number of items, each of which has an independently defined type.
+/// </summary>
+public class TupleType : ArrayType
+{
+    private readonly Lazy<ITypeReference> lazyItem;
+
+    public TupleType(string name, ImmutableArray<ITypeReference> items, TypeSymbolValidationFlags validationFlags) : base(name)
+    {
+        Items = items;
+        ValidationFlags = validationFlags;
+        // Items may contain DeferredTypeReferences, so only calculate Item on request
+        lazyItem = new(() => TypeHelper.CreateTypeUnion(this.Items));
+    }
+
+    public override TypeSymbolValidationFlags ValidationFlags { get; }
+
+    public ImmutableArray<ITypeReference> Items { get; }
+
+    public override ITypeReference Item => lazyItem.Value;
+
+    public override bool Equals(object? other) =>
+        other is TupleType otherTuple ? otherTuple == this : false;
+
+    public override int GetHashCode() => (GetType(), Items, ValidationFlags).GetHashCode();
+
+    public static bool operator ==(TupleType? a, TupleType? b) => a is not null
+        ? (b is not null ? nonNullEquals(a, b) : false)
+        : b is null;
+
+    public static bool operator !=(TupleType? a, TupleType? b) => !(a == b);
+
+    private static bool nonNullEquals(TupleType a, TupleType b) => a.ValidationFlags == b.ValidationFlags &&
+        a.Items.Length == b.Items.Length &&
+        Enumerable.Range(0, a.Items.Length).All(idx => a.Items[idx].Equals(b.Items[idx]));
```

Comment:
I think it would be worth aligning with other implementations we have in this codebase. For `Equals()` & `GetHashCode()`:
https://github.com/Azure/bicep/blob/4543562544af009134d8528fe26ca08b4277037f/src/Bicep.Core/Resources/ResourceTypeReference.cs#L86-L100

For operator overloading `==` & `!=` - I'm keen on inverting this (`==` calls `Equals()`). For example, the .NET `string` implementation:
https://github.com/microsoft/referencesource/blob/dae14279dd0672adead5de00ac8f117dcf74c184/mscorlib/system/string.cs#L705-L711

Reply:
Also, I'm personally of the opinion that operator overloading generally is best avoided unless there's a really good reason for it - for maintainability reasons (it's so rarely done, that other readers of the code generally assume it's a reference equality comparison). Are you sure we need it?

Reply:
> Are you sure we need it?

Not at all! I wasn't sure if this was standard in C# and am happy to ditch it if it's not expected to be defined alongside `Equals` and `GetHashCode`

Reply:
> Not at all! I wasn't sure if this was standard in C# and am happy to ditch it if it's not expected to be defined alongside `Equals` and `GetHashCode`

Sounds good!